### ISSUE TITLE
In a tablebase position that is winning without the 50 moves rule, Stockfish gives quickest winning move as the only move that draws

### ISSUE THREAD

**Issue by bennedik** [official-stockfish/Stockfish#828]
Example FEN: 
r7/8/2N5/2N2k1b/3K4/8/8/8 b - - 0 1
This is from a game Karjakin-Shirov, 2007.

Actual behaviour:
Stockfish gives 1..Re8 as the only move that draws.
Without 50 moves rule, 1..Re8 is the quickest win (mate in 178).
No other moves are listed in the search results.

Expected behaviour:
Any move draws.
More than one move should be listed in Stockfish's search results.

**Comment by Stefano80** [comment]
>Expected behaviour

Expected by whom? If I understand correctly, SF wants to play 1. ... Re8 and this happens to be the best move. So, what is the issue exactly?

**Comment by bennedik** [comment]
> Expected by whom? If I understand correctly, SF wants to play 1. ... Re8 and this happens to be the best move. So, what is the issue exactly?

You are right, that Stockfish does play the best move.
However, Stockfish is not only used for playing, but also for analysis.
Probably a lot of chess players don't play against Stockfish, but use it to help with their analysis.

For me personally, that is even more important than playing, because I am using Stockfish within my algorithm to create chess puzzles, see https://chesspuzzle.net.

Usually Stockfish would list all the drawing moves, not just one.
For example:
4k3/8/4P3/8/5K2/8/8/8 b - - 0 1
Here Stockfish gives all the moves Kf8, Ke7 and Kd8 as draw.

From the analysis in the Karjakin-Shirov position, I can't see that Re8 is not the only drawing move.
The analysis looks like all other moves are lost.
Is there any way this can be improved?
I am sure it would be useful not only for my algorithm, but also for anyone who uses Stockfish for analysis.

**Comment by ddugovic** [comment]
When I use 5-man Syzygybases and set MultiPV to 2 I fail to duplicate the issue (which I don't perceive to be a problem for analysis since you can specify candidate moves in the `go` command and thereby force SF to select from that list of moves):

```
info depth 20 seldepth 34 multipv 1 score cp 56 nodes 3498742 nps 1838540 hashfull 916 tbhits 121883 time 1903 pv h5f3 c6e7 f5g4 c5d3 g4g5 d3e5 a8a4 d4c5 a4a5 c5d4 a5a4 d4c5 a4a5 c5d4 f3b7 e7c6 a5a4 d4c5 a4a1 e5d3 a1h1 c6e5 h1h5 c5d6 h5h1 d3c5 h1h6 d6c7
info depth 20 seldepth 34 multipv 2 score cp 54 nodes 3498742 nps 1838540 hashfull 916 tbhits 121883 time 1903 pv a8e8 d4d5 h5f7 d5d6 f5f4 c5d3 f4e3 d3e5 f7a2 c6e7 e8d8 e5d7 e3f4 e7g6 f4g5 g6e7 d8h8 d7c5 g5g4 c5e4 h8h6 d6c5 h6h5 c5b4 g4f4 e4f6
bestmove h5f3 ponder c6e7
```

**Comment by syzygy1** [comment]
Multipv indeed does not work as one might expect for TB positions.

The TB root probe code filters out all "inferior" moves. Stockfish then performs a search on the remaining moves. In principle this is fine: it makes no sense to consider the inferior moves (which could cost a half or whole point) and the search will normally pick the most sensible "optimal" move. But unfortunately, filtering out the inferior moves may lead to the display of fewer PV lines than what multipv is set to.

Fixing this is not trivial. If SF does not filter out the inferior moves, the search might find one of those moves to be better than the optimal moves and show it on top. During analysis this would not lose points (as no game is being played), but it would certainly greatly confuse people.

What SF would need to do is this:
- instead of filtering out bad root moves in a TB position, it should "score" all root moves.
- the multipv "for" loop in thread_search() should first search the set of root moves with a top score.
- If there are at least as many of those as the multipv setting requires, then all is fine.
- If not, another multipv "for" loop needs to be executed on the next batch of root moves (e.g. on all the drawing moves).
- If we now have enough PV lines, we are done.
- If not, we need to execute yet another loop on the next batch of root moves (e.g. the losing moves).

I have not implemented this, but I think it should work. It will also search the UCI "searchmoves" option, which is also somewhat broken now. Any conceptually simpler approach runs into problems even if they are not immediately apparent.

In short, fixing this will add a number of lines to Stockfish but it will not add Elo. A solution is therefore rather unlikely to make it into Stockfish.

However, I was anyway planning to fix this in Cfish. When that is done, at least the code is available.

**Comment by syzygy1** [comment]
@ddugovic 
Current SF will filter out those (suboptimal) candidate moves, I am afraid :)

**Comment by bennedik** [comment]
@ddugovic I am using 6 man tablebases. It can probably not be reproduced with 5 man.

**Comment by bennedik** [comment]
@syzygy1 Thank you for your detailed answer.

In the case of a table base position without the 50 move rule interfering, it is not a problem at all for my puzzle generation algorithm, that Stockfish filters out the non-optimal moves.
As long as I get all the winning moves in a winning position, and all the drawing moves in a drawing position there is no issue.
But in this case it looks as if Re8 is the only drawing move, while in fact any other move draws.
(The algorithm constructed an amusing drawing puzzle with a long sequence of only-moves...)

Is there any way to recognize that this is a case for the 50 move rule?
I will try the same puzzle with the Syzygy50MoveRule parameter set to false.
Maybe I can recognize this special case by comparing the result of calculating the position with the parameter turned on and off.

By the way, is there any meaning in the evaluation score in table base evaluations?
For example does 12353 mean something else than 12345?

**Comment by syzygy1** [comment]
Syzygy50Move won't make a difference here.
It makes sense to consider only the sole "cursed" winning move when selecting a move to play. Although that move will draw against optimal play, it will normally have the best winning chances against suboptimal play.
The only problem is the interaction with multipv.

Note that you can find the whole list of moves plus outcomes on web sites like:
https://syzygy-tables.info/?fen=r7/8/2N5/2N2k1b/3K4/8/8/8_b_-_-_0_1
As you see there are in fact many moves that "win" but all outside the 50-move rule.
Stockfish's TB root probing selects only the "best" one (with minimal DTZ value) since that is probably the best black can do here. If white's play is sufficiently inaccurate, black might manage to get DTZ to a value within the remaining 50-move budget before the game is drawn.

Given the availability of such web sites, the value of "fixing" this in Stockfish (at the cost of added complexity) is a bit questionable, I am afraid. But I will make an effort in my own fork.

**Comment by syzygy1** [comment]
The TB winning scores have no real meaning. But the higher they are, the closer the position is to a TB win. So if Stockfish initially shows a TB win with say 128.30 (123.xx is 128.xx nowadays due to an internal change in SF), then a few moves later the score might have become 128.40. This score will continue to increase until an actual 6-piece position is on the board.

**Comment by bennedik** [comment]
@syzygy1 I have now tested this with the two different values for the Syzygy50MoveRule parameter:

setoption name MultiPV value 10
go depth 1
info depth 1 seldepth 1 multipv 1 score cp 0 nodes 499 nps 166333 tbhits 1 time 3 pv a8e8
bestmove a8e8
setoption name Syzygy50MoveRule value false
go depth 1
info depth 1 seldepth 1 multipv 1 score cp 12353 nodes 499 nps 249500 tbhits 1 time 2 pv a8e8
bestmove a8e8

You are right, there is no difference in the number of lines displayed.
However, there is a difference in the evaluation.

I will add a validation to my puzzle algorithm: If the position is a table base position, and the evaluation differs when changing the Syzygy50MoveRule parameter, I will just discard the puzzle.

I think there is never going to be an interesting puzzle in such positions anyway. Or the solution is interesting only because it is so long - but no human would be able to solve it.

I am closing this issue, thank you for your explanations.
